src/Workspaces/Core/Portable/FindSymbols/Shared/AbstractSyntaxIndex.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal abstract partial class AbstractSyntaxIndex<TIndex>
```

Comment:
this is all pulled up code from SyntaxTreeIndex so that we can share a lot of loading/persisting code with SyntaxTreeIndex and TopLevelSyntaxTreeIndex.  It is effectively unchanged except that it now takes delegates to the actual reading/writing functions which the subclasses provide.